### PR TITLE
[IMP] mail: add attachment panel to discuss

### PR DESCRIPTION
**Pull request by tsm-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/static/src/discuss/core/common/thread_actions.js [hunk 1]
```diff
@@ -10,6 +11,18 @@ import { _t } from "@web/core/l10n/translation";
...
+            component.thread?.model === "discuss.channel" &&
+            (!component.props.chatWindow || component.props.chatWindow.isOpen),
+        component: AttachmentPanel,
+        icon: "fa fa-fw fa-file",
+        iconLarge: "fa fa-fw fa-lg fa-file",
+        name: _t("Show Attachment List"),
+        nameActive: _t("Hide Attachment List"),
+        sequence: 25,
+        toggle: true,
+    })
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/thread_actions.js:25]: Not something to change in this PR, just pointing out an issue with some thread actions like this one:
The reliance on component props and getters is not good, because these thread actions have prerequisite on the component that can easily omitted, thus making them not working.
Also they might conflict with existing props that mean something (e.g. getter `thread` doesn't necessarily mean the thread being viewed, e.g. `Composer@thread` is the origin thread).

I'm still thinking on a better solution. We could use `setup()` in the meantime to document what it's expected for the component to have, and what it means like the getter `thread` here.

File: addons/mail/static/src/discuss/core/common/attachment_panel.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-name="mail.AttachmentPanel">
+        <t t-log="attachmentsByDate"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.xml:5]: `t-log` to remove

File: addons/mail/static/src/discuss/core/common/attachment_panel.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        return _t("Attachments");
+    }
+
+    get emptyMessage() {
+        if (this.props.thread.type === "channel") {
+            return _t("This channel doesn't have any attachments.");
+        } else {
+            return _t("This conversation doesn't have any attachments.");
+        }
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.js:49]: This getter is not worth it. Better put branching immediately in the template

File: addons/mail/static/src/discuss/core/common/attachment_panel.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        const loadOlderState = useVisible("load-older", () => {
+            if (loadOlderState.isVisible) {
+                this.threadService.fetchMoreAttachments(this.props.thread);
+            }
+        });
+    }
+
+    get title() {
+        return _t("Attachments");
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.js:41]: I think with recent improved `_t()`, you can just `<t t-set t-value/>` in template.
This is one less indirection if this works

File: addons/mail/static/tests/chat_window/chat_window_tests.js [hunk 5]
```diff
@@ -148,7 +148,7 @@ QUnit.test("chat window: basic rendering", async (assert) => {
         "There are no messages in this conversation."
     );
     await click("[title='Open Actions Menu']");
-    await contains(".o-mail-ChatWindow-command", 9);
+    await contains(".o-mail-ChatWindow-command", 10);
```
Comment (alexkuhn) [inline addons/mail/static/tests/chat_window/chat_window_tests.js:151]: You should add the newly added `contains()`.

This tests asserts `4` then enumerate them. Then asserts `9` and enumerates the 5 extras.
Since counter is now 10, it should enumerate 6 extras (+1 with attachment)

File: addons/mail/static/tests/discuss/core/attachment_panel_tests.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    const pyEnv = await startServer();
+    const channelId = await pyEnv["discuss.channel"].create({ name: "General" });
+    const { openDiscuss } = await start();
+    await openDiscuss(channelId);
+
+    await click(".o-mail-Discuss-header button[title='Show Attachment List']");
+    await contains(
+        ".o-mail-Discuss-inspector:contains(This channel doesn't have any attachments.)"
+    );
+});
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/core/attachment_panel_tests.js:17]: - empty newline is useless. Better remove it so it's clearer at a glance these LOCs are grouped together for a single test
- Name of test is misleading. "message" is used a lot to refer to an actual `mail.message`. I would reword to "Empty attachment panel"

File: addons/mail/static/tests/discuss/core/attachment_panel_tests.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+QUnit.test("Attachment panel display attachments sorted by date", async () => {
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/core/attachment_panel_tests.js:19]: `Attachment panel sort by date` is more brief (full sentence gives impression test is more complex than it is)

File: addons/mail/static/src/discuss/core/common/attachment_panel.js (hunk 4 above)
Comment (oomsveta) [reply addons/mail/static/src/discuss/core/common/attachment_panel.js]: AFAIK `t-value` is not exported, but you can do something like this:
```xml
<t t-set="title">Attachments</t>
```

File: addons/mail/static/src/discuss/core/common/thread_actions.js [hunk 8]
```diff
@@ -10,6 +11,18 @@ import { _t } from "@web/core/l10n/translation";
...
 threadActionsRegistry
+    .add("attachments", {
+        condition: (component) =>
+            component.thread?.model === "discuss.channel" &&
+            (!component.props.chatWindow || component.props.chatWindow.isOpen),
+        component: AttachmentPanel,
+        icon: "fa fa-fw fa-paperclip",
+        iconLarge: "fa fa-fw fa-lg fa-paperclip",
+        name: _t("Show Attachment List"),
+        nameActive: _t("Hide Attachment List"),
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/thread_actions.js:22]: Nit-picking: _"Show/Hide Attachments"_ is more brief and to the point. "List" adds another word, makes it less readable overall and has no improvement over plural "Attachments"

File: addons/mail/static/src/discuss/core/common/attachment_panel.xml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+    <t t-name="mail.AttachmentPanel">
+        <t t-set="title">Attachments</t>
+        <ActionPanel title="title">
+            <div class="flex-grow-1" t-att-class="{
+                'd-flex justify-content-center align-items-center': props.thread.attachments.length === 0,
+            }">
+                <p t-if="props.thread.attachments.length === 0" class="text-center fst-italic text-500 fs-6">
+                    <t t-if="props.thread.type === 'channel'">This channel doesn't have any attachments.</t>
+                    <t t-else="">This conversation doesn't have any attachments.</t>
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.xml:12]: Note that `AttachmentPanel` only used with channels. The `t-else` is currently unreachable

File: addons/mail/static/src/core/common/attachment_model.js [hunk 10]
```diff
@@ -54,6 +57,11 @@ export class Attachment extends Record {
         return this.mimetype && this.mimetype.startsWith("application/pdf");
     }
 
+    get dateString() {
+        const datetime = deserializeDateTime(this.create_date);
+        return `${datetime.monthLong}, ${datetime.year}`;
+    }
+
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/attachment_model.js:64]: Name `dateString` is a bit obscure in the actual formatting of string. What about `monthYear`? "String" doesn't had anything, let the typing tell it's a string, rather than put it in name of getter

File: addons/mail/static/src/discuss/core/common/attachment_panel.xml (hunk 9 above)
Comment (tsm-odoo) [reply addons/mail/static/src/discuss/core/common/attachment_panel.xml:12]: It is reachable when in a 1-to-1 chat

**Comment by tsm-odoo** [comment]
@alexkuhn done :+1:

File: addons/mail/static/src/discuss/core/common/attachment_panel.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    static template = "mail.AttachmentPanel";
+
+    setup() {
+        this.threadService = useService("mail.thread");
+        this.attachmentUploadService = useService("mail.attachment_upload");
+        onWillStart(() => {
+            this.threadService.fetchMoreAttachments(this.props.thread);
+        });
+        onWillUpdateProps((nextProps) => {
+            if (nextProps.thread.localId !== this.props.thread.localId) {
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.js:27]: Could use `nextProps.thread.notEq(this.props.thread)`, rather than compare on `localId`. This is cleaner

File: addons/mail/static/src/discuss/core/common/attachment_panel.xml (hunk 9 above)
Comment (alexkuhn) [reply addons/mail/static/src/discuss/core/common/attachment_panel.xml:12]: Oops, misread `.model === 'discuss.channel'`. Then it's ok

File: addons/mail/static/src/discuss/core/common/attachment_panel.xml [hunk 12]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-name="mail.AttachmentPanel">
+        <t t-set="title">Attachments</t>
+        <ActionPanel title="title">
+            <div class="flex-grow-1" t-att-class="{
+                'd-flex justify-content-center align-items-center': props.thread.attachments.length === 0,
+            }">
+                <p t-if="props.thread.attachments.length === 0" class="text-center fst-italic text-500 fs-6">
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.xml:10]: Haven't tested functionally. Not entirely sure `fs-6` is necessary (this is the default font-size)

**Review by alexkuhn — COMMENTED** [review]
some other comments

File: addons/mail/static/src/discuss/core/common/thread_service_patch.js [hunk 13]
```diff
@@ -27,4 +27,26 @@ patch(ThreadService.prototype, {
...
+        thread.isLoadingAttachments = true;
+        try {
+            const rawAttachments = await this.rpc("/discuss/channel/attachments", {
+                before: Math.min(...thread.attachments.map(({ id }) => id)),
+                channel_id: thread.id,
+                limit,
+            });
+            const attachments = rawAttachments.map((rawAttachment) =>
+                this.store.Attachment.insert(rawAttachment)
+            );
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/thread_service_patch.js:44]: Maybe a short name for variable in loop could be single line and make prettier is happy?
This is a case where less LOCs can be beneficial thanks to shorter names. Even a plain `data` or `d` is good, it's just a variable in this tiny scope

File: addons/mail/static/tests/discuss/core/attachment_panel_tests.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+            create_date: "2023-09-21",
+        },
+    ]);
+    const { openDiscuss } = await start();
+    await openDiscuss(channelId);
+    await click(".o-mail-Discuss-header button[title='Show Attachment List']");
+    await contains(
+        ".o-mail-DateSeparator:contains(August, 2023) + .o-mail-AttachmentList:contains(file1.pdf)"
+    );
+    await contains(
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/core/attachment_panel_tests.js:41]: Don't forget adapt test for "Show Attachments" string changed

File: addons/mail/static/src/discuss/core/common/attachment_panel.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        const loadOlderState = useVisible("load-older", () => {
+            if (loadOlderState.isVisible) {
+                this.threadService.fetchMoreAttachments(this.props.thread);
+            }
+        });
+    }
+
+    /**
+     * @return {Object<string, import("@mail/model/model").Attachment[]>}
+     */
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/core/common/attachment_panel.js:40]: `@mail/model/model` looks incorrect path

File: addons/mail/static/src/core/common/attachment_model.js [hunk 16]
```diff
@@ -54,6 +57,11 @@ export class Attachment extends Record {
         return this.mimetype && this.mimetype.startsWith("application/pdf");
     }
 
+    get monthYear() {
+        const datetime = deserializeDateTime(this.create_date);
+        return `${datetime.monthLong}, ${datetime.year}`;
+    }
+
```
Comment (alexkuhn) [inline addons/mail/static/src/core/common/attachment_model.js:64]: I'm not 100% all cases of inserting an Attachment pass `create_date`.
I would play safe and let it return `undefined` if there's no `create_date`, and let `AttachmentPanel` show these attachments in their own section if this happens in rare cases.

**Review by alexkuhn — COMMENTED** [review]
Code looks good to me otherwise!

**Comment by tsm-odoo** [comment]
@alexkuhn I think we're ready to go then :+1:

File: addons/mail/static/tests/chat_window/chat_window_tests.js [hunk 17]
```diff
@@ -148,8 +148,9 @@ QUnit.test("chat window: basic rendering", async (assert) => {
...
     await contains("[title='Pinned Messages']");
+    await contains("[title='Show Attachment List']");
```
Comment (alexkuhn) [inline addons/mail/static/tests/chat_window/chat_window_tests.js:153]: `Show Attachments`

**Review by alexkuhn — APPROVED** [review]
LGTM otherwise
@robodoo delegate+

**Comment by tsm-odoo** [comment]
@robodoo r+

**Comment by tsm-odoo** [comment]
@robodoo r+
